<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Descriptions of available values for 'columns' and...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for UNIPROTKB {UniProt.ws}"><tr><td>UNIPROTKB {UniProt.ws}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Descriptions of available values for <code>columns</code> and <code>keytypes</code>.</h2>

<h3>Description</h3>

<p>This manual page enumerates the kinds of data represented by the
values returned when the user calls <code>columns</code> or <code>keytypes</code>
</p>


<h3>Details</h3>

  
<p>All the possible values for <code>columns</code> and <code>keytypes</code> are listed
below.  Users will have to actually use these methods to learn which
of the following possible values actually apply in their case.
</p>

<dl>
<dt>:UNIPROTKB</dt><dd><p>The central ID for UniProt and swissprot</p>
</dd>
<dt>:UNIPARC</dt><dd><p>UniParc</p>
</dd>
<dt>:UNIREF50</dt><dd><p>UniRef50</p>
</dd>
<dt>:UNIREF90</dt><dd><p>UniRef90</p>
</dd>
<dt>:UNIREF100</dt><dd><p>UniRef100</p>
</dd>
<dt>:EMBL/GENBANK/DDBJ</dt><dd><p>EMBL/GenBank/DDBJ</p>
</dd>
<dt>:EMBL/GENBANK/DDBJ_CDS</dt><dd><p>EMBL/GenBank/DDBJ CDS</p>
</dd>
<dt>:PIR</dt><dd><p>PIR</p>
</dd>
<dt>:UNIGENE</dt><dd><p>UniGene</p>
</dd>
<dt>:ENTREZ_GENE</dt><dd><p>Entrez Gene (GeneID)</p>
</dd>
<dt>:GI_NUMBER*</dt><dd><p>GI number</p>
</dd>
<dt>:IPI</dt><dd><p>IPI</p>
</dd>
<dt>:REFSEQ_PROTEIN</dt><dd><p>RefSeq Protein</p>
</dd>
<dt>:REFSEQ_NUCLEOTIDE</dt><dd><p>RefSeq Nucleotide</p>
</dd>
<dt>:PDB</dt><dd><p>PDB</p>
</dd>
<dt>:DISPROT</dt><dd><p>DisProt</p>
</dd>
<dt>:HSSP</dt><dd><p>HSSP</p>
</dd>
<dt>:DIP</dt><dd><p>DIP</p>
</dd>
<dt>:MINT</dt><dd><p>MINT</p>
</dd>
<dt>:ALLERGOME</dt><dd><p>Allergome</p>
</dd>
<dt>:MEROPS</dt><dd><p>MEROPS</p>
</dd>
<dt>:PEROXIBASE</dt><dd><p>PeroxiBase</p>
</dd>
<dt>:PPTASEDB</dt><dd><p>PptaseDB</p>
</dd>
<dt>:REBASE</dt><dd><p>REBASE</p>
</dd>
<dt>:TCDB</dt><dd><p>TCDB</p>
</dd>
<dt>:PHOSSITE</dt><dd><p>PhosSite</p>
</dd>
<dt>:DMDM</dt><dd><p>DMDM</p>
</dd>
<dt>:AARHUS/GHENT-2DPAGE</dt><dd><p>Aarhus/Ghent-2DPAGE</p>
</dd>
<dt>:ECO2DBASE</dt><dd><p>ECO2DBASE</p>
</dd>
<dt>:WORLD-2DPAGE</dt><dd><p>World-2DPAGE</p>
</dd>
<dt>:DNASU</dt><dd><p>DNASU</p>
</dd>
<dt>:ENSEMBL</dt><dd><p>Ensembl</p>
</dd>
<dt>:ENSEMBL_PROTEIN</dt><dd><p>Ensembl Protein</p>
</dd>
<dt>:ENSEMBL_TRANSCRIPT</dt><dd><p>Ensembl Transcript</p>
</dd>
<dt>:ENSEMBL_GENOMES</dt><dd><p>Ensembl Genomes</p>
</dd>
<dt>:ENSEMBL_GENOMES PROTEIN</dt><dd><p>Ensembl Genomes Protein</p>
</dd>
<dt>:ENSEMBL_GENOMES TRANSCRIPT</dt><dd><p>Ensembl Genomes Transcript</p>
</dd>
</dl>


<dl>
<dt>:KEGG</dt><dd><p>KEGG</p>
</dd>
<dt>:PATRIC</dt><dd><p>PATRIC</p>
</dd>
<dt>:TIGR</dt><dd><p>TIGR</p>
</dd>
<dt>:UCSC</dt><dd><p>UCSC</p>
</dd>
<dt>:VECTORBASE</dt><dd><p>VectorBase</p>
</dd>
<dt>:AGD</dt><dd><p>AGD</p>
</dd>
<dt>:ARACHNOSERVER</dt><dd><p>ArachnoServer</p>
</dd>
<dt>:CGD</dt><dd><p>CGD</p>
</dd>
<dt>:CONOSERVER</dt><dd><p>ConoServer</p>
</dd>
<dt>:CYGD</dt><dd><p>CYGD</p>
</dd>
<dt>:DICTYBASE</dt><dd><p>dictyBase</p>
</dd>
<dt>:ECHOBASE</dt><dd><p>EchoBASE</p>
</dd>
<dt>:ECOGENE</dt><dd><p>EcoGene</p>
</dd>
<dt>:EUHCVDB</dt><dd><p>euHCVdb</p>
</dd>
<dt>:EUPATHDB</dt><dd><p>EuPathDB</p>
</dd>
<dt>:FLYBASE</dt><dd><p>FlyBase</p>
</dd>
<dt>:GENECARDS</dt><dd><p>GeneCards</p>
</dd>
<dt>:GENEFARM</dt><dd><p>GeneFarm</p>
</dd>
<dt>:GENOLIST</dt><dd><p>GenoList</p>
</dd>
<dt>:H-INVDB</dt><dd><p>H-InvDB</p>
</dd>
<dt>:HGNC</dt><dd><p>HGNC</p>
</dd>
<dt>:HPA</dt><dd><p>HPA</p>
</dd>
<dt>:LEGIOLIST</dt><dd><p>LegioList</p>
</dd>
<dt>:LEPROMA</dt><dd><p>Leproma</p>
</dd>
<dt>:MAIZEGDB</dt><dd><p>MaizeGDB</p>
</dd>
<dt>:MIM</dt><dd><p>MIM</p>
</dd>
<dt>:MGI</dt><dd><p>MGI</p>
</dd>
<dt>:NEXTPROT</dt><dd><p>neXtProt</p>
</dd>
<dt>:ORPHANET</dt><dd><p>Orphanet</p>
</dd>
<dt>:PHARMGKB</dt><dd><p>PharmGKB</p>
</dd>
<dt>:POMBASE</dt><dd><p>PomBase</p>
</dd>
<dt>:PSEUDOCAP</dt><dd><p>PseudoCAP</p>
</dd>
<dt>:RGD</dt><dd><p>RGD</p>
</dd>
<dt>:SGD</dt><dd><p>SGD</p>
</dd>
<dt>:TAIR</dt><dd><p>TAIR</p>
</dd>
<dt>:TUBERCULIST</dt><dd><p>TubercuList</p>
</dd>
<dt>:WORMBASE</dt><dd><p>WormBase</p>
</dd>
<dt>:WORMBASE_TRANSCRIPT</dt><dd><p>WormBase Transcript</p>
</dd>
<dt>:WORMBASE_PROTEIN</dt><dd><p>WormBase Protein</p>
</dd>
<dt>:XENBASE</dt><dd><p>Xenbase</p>
</dd>
<dt>:ZFIN</dt><dd><p>ZFIN</p>
</dd>
<dt>:EGGNOG</dt><dd><p>eggNOG</p>
</dd>
<dt>:GENETREE</dt><dd><p>GeneTree</p>
</dd>
<dt>:HOVERGEN</dt><dd><p>HOVERGEN</p>
</dd>
<dt>:KO</dt><dd><p>KO</p>
</dd>
<dt>:OMA</dt><dd><p>OMA</p>
</dd>
<dt>:ORTHODB</dt><dd><p>OrthoDB</p>
</dd>
<dt>:PROTCLUSTDB</dt><dd><p>ProtClustDB</p>
</dd>
<dt>:BIOCYC</dt><dd><p>BioCyc</p>
</dd>
<dt>:REACTOME</dt><dd><p>Reactome</p>
</dd>
<dt>:UNIPATHWAY</dt><dd><p>UniPathWay</p>
</dd>
<dt>:CLEANEX</dt><dd><p>CleanEx</p>
</dd>
<dt>:GERMONLINE</dt><dd><p>GermOnline</p>
</dd>
<dt>:DRUGBANK</dt><dd><p>DrugBank</p>
</dd>
<dt>:GENOMERNAI</dt><dd><p>GenomeRNAi</p>
</dd>
<dt>:NEXTBIO</dt><dd><p>NextBio</p>
</dd>
<dt>:CITATION</dt><dd><p>citations</p>
</dd>
<dt>:CLUSTERS</dt><dd><p>clusters</p>
</dd>
<dt>:COMMENTS</dt><dd><p>comments</p>
</dd>
<dt>:DOMAINS</dt><dd><p>domains</p>
</dd>
<dt>:DOMAIN</dt><dd><p>domain</p>
</dd>
<dt>:EC</dt><dd><p>ec ID</p>
</dd>
<dt>:ID</dt><dd><p>ID</p>
</dd>
</dl>

<dl>
<dt>:EXISTENCE</dt><dd><p>existence</p>
</dd>
<dt>:FAMILIES</dt><dd><p>families</p>
</dd>
<dt>:FEATURES</dt><dd><p>features</p>
</dd>
<dt>:GENES</dt><dd><p>genes</p>
</dd>
<dt>:GO</dt><dd><p>go term</p>
</dd>
<dt>:GO-ID</dt><dd><p>go id</p>
</dd>
<dt>:INTERPRO</dt><dd><p>interpro</p>
</dd>
<dt>:INTERACTOR</dt><dd><p>interactor</p>
</dd>
<dt>:KEYWORDS</dt><dd><p>keywords</p>
</dd>
<dt>:KEYWORD-ID</dt><dd><p>keyword-id</p>
</dd>
<dt>:LAST-MODIFIED</dt><dd><p>last-modified</p>
</dd>
<dt>:LENGTH</dt><dd><p>length</p>
</dd>
<dt>:ORGANISM</dt><dd><p>organism</p>
</dd>
<dt>:ORGANISM-ID</dt><dd><p>organism-id</p>
</dd>
<dt>:PATHWAY</dt><dd><p>pathway</p>
</dd>
<dt>:PROTEIN NAMES</dt><dd><p>protein names</p>
</dd>
<dt>:REVIEWED</dt><dd><p>reviewed</p>
</dd>
<dt>:SCORE</dt><dd><p>score</p>
</dd>
<dt>:SEQUENCE</dt><dd><p>sequence</p>
</dd>
<dt>:3D</dt><dd><p>3d</p>
</dd>
</dl>

<dl>
<dt>:TAXON</dt><dd><p>taxon</p>
</dd>
<dt>:TOOLS</dt><dd><p>tools</p>
</dd>
<dt>:VERSION</dt><dd><p>version</p>
</dd>
</dl>

<dl>
<dt>:DATABASE(PFAM)</dt><dd><p>PFAM ids</p>
</dd>
<dt>:DATABASE(PDB)</dt><dd><p>PDB ids</p>
</dd>
<dt>:</dt><dd></dd>
</dl>



<h3>Author(s)</h3>

<p>Marc Carlson</p>


<h3>Examples</h3>

<pre>
  up &lt;- UniProt.ws(taxId=9606)
  ## List the possible values for columns
  columns(up)
  ## List the possible values for keytypes
  keytypes(up)
  ## get some values back
  ## list all possible keys of type entrez gene ID.
  ## (this process is not instantaneous)
  if(interactive()){
    keys &lt;- head(keys(up, keytype="UNIPROTKB"))
    keys
  }
  select(up, keys=c("P31946","P62258"), columns=c("PDB","SEQUENCE"),
  keytype="UNIPROTKB")
</pre>

<hr /><div style="text-align: center;">[Package <em>UniProt.ws</em> version 2.22.0 <a href="00Index.html">Index</a>]</div>
</body></html>
